### ISSUE TITLE
Mechanism to pass additional attributes to custom eval or loss functions

### ISSUE THREAD

**Issue by jayantsahewal** [microsoft/LightGBM#4009]
Hello Team,

Great work on the package. I know there is a support for custom eval and loss function which is awesome. However, is there a way to pass additional attribute(s) (maybe part of feature set or maybe not). For example, imagine if there is a certain attribute by which we want the parity in the model's performance, then mathematically it's easy to create a custom eval function, but the function will need more data than just the predictions and labels. So, the question is, does there such mechanism exist in LightGBM?

Best,
Jayant

**Comment by jameslamb** [comment]
Thanks for using LightGBM and opening this issue.

Could you be more specific about what types of parameters you'd like to pass? We might be able to suggest a way to accomplish what you'd like to accomplish without needing to change the way `lightgbm` handles custom metric / objective functions.

**Comment by jayantsahewal** [comment]
Yes, definitely. For example, let's say my training data has a column/feature which take values 'A', 'B', and 'C' and I want to make sure that the False Positive Rate on the evaluation data (which I use for early stopping criteria) is same across all values that feature can take.

**Comment by StrikerRUS** [comment]
@jayantsahewal I think you can use [`train()` function](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html#lightgbm.train) from standard API. `fobj` and `feval` arguments of that function are callables which take whole Dataset object via `train_data` argument. So you can compute everything you want from original features. Just don't forget to specify `free_raw_data=False` during Dataset creation.

**Comment by shiyu1994** [comment]
@StrikerRUS I think we cannot get the origin feature values from `train_data` in python directly? A dirty work around would be like this, which explicitly record the original feature value in python and directly use it in the evaluation function.
```python
X = np.array([[1, 2, 3], [4, 5, 6]])
y = np.array([0, 1])
train_data = lgb.Dataset(data=X, label=y)

def custom_eval_func(preds, train_data):
    return "nonsense", np.mean(X[:, 0] + preds), True

params = {'objective':'regression'}
lgb.train(params=params, train_set=train_data, valid_sets=[train_data], valid_names=["train"], feval=custom_eval_func)
```

**Comment by StrikerRUS** [comment]
@shiyu1994 I believe we can. Just as  I mentioned above, `free_raw_data` should be `False`. Please consider checking the following snippet (reworked your code):

```py
import lightgbm as lgb
import numpy as np

X = np.array([[1, 2, 3], [4, 5, 6]])
y = np.array([0, 1])
train_data = lgb.Dataset(data=X, label=y, free_raw_data=False)

def custom_eval_func(preds, train_data):
    return "nonsense", np.mean(train_data.get_data()[:, 0] + preds), True

params = {'objective': 'regression'}
lgb.train(params=params, train_set=train_data, valid_sets=[train_data], valid_names=["train"], feval=custom_eval_func)
```

**Comment by shiyu1994** [comment]
Sorry. I did not noticed the `get_data()` method.

**Comment by StrikerRUS** [comment]
@shiyu1994 No problem! `Dataset` class is really "fat", it is impossible to remember all methods.

@jayantsahewal Does [`get_data()`](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Dataset.html#lightgbm.Dataset.get_data) method suit your needs?
